packages/@aws-cdk/core/lib/private/refs.ts

```diff
@@ -0,0 +1,276 @@
+// ----------------------------------------------------
+// CROSS REFERENCES
+// ----------------------------------------------------
+import { CfnElement } from '../cfn-element';
+import { CfnOutput } from '../cfn-output';
+import { CfnParameter } from '../cfn-parameter';
+import { Construct } from '../construct-compat';
+import { Reference } from '../reference';
+import { IResolvable } from '../resolvable';
+import { Stack } from '../stack';
+import { Token } from '../token';
+import { CfnReference } from './cfn-reference';
+import { Intrinsic } from './intrinsic';
+import { findTokens } from './resolve';
+import { makeUniqueId } from './uniqueid';
+
+/**
+ * This is called from the App level to resolve all references defined. Each
+ * reference is resolved based on it's consumption context.
+ */
+export function resolveReferences(scope: Construct) {
+  const edges = findAllReferences(scope);
+
+  for (const { source, value } of edges) {
+    const consumer = Stack.of(source);
+
+    // skip if we already have a value for this consumer
+    if (value.hasValueForStack(consumer)) {
+      continue;
+    }
+
+    // resolve the value in the context of the consumer.
+    const resolved = resolveValue(consumer, value);
+    value.assignValueForStack(consumer, resolved);
+  }
+}
+
+/**
+ * Resolves the value for `reference` in the context of `consumer`.
+ */
+function resolveValue(consumer: Stack, reference: CfnReference): IResolvable {
+  const producer = Stack.of(reference.target);
+
+  // produce and consumer stacks are the same, we can just return the value itself.
+  if (producer === consumer) {
+    return reference;
+  }
+
+  // unsupported: stacks from different apps
+  if (producer.node.root !== consumer.node.root) {
+    throw new Error('Cannot reference across apps. Consuming and producing stacks must be defined within the same CDK app.');
+  }
+
+  // unsupported: stacks are not in the same environment
+  if (producer.environment !== consumer.environment) {
+    throw new Error(
+      `Stack "${consumer.node.path}" cannot consume a cross reference from stack "${producer.node.path}". ` +
+      'Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack');
+  }
+
+  // ----------------------------------------------------------------------
+  // consumer is nested in the producer (directly or indirectly)
+  // ----------------------------------------------------------------------
+
+  // if the consumer is a child of the producer, wire the reference through a
+  // CloudFormation parameter on the consumer and resolve recursively.
+  if (isParent(producer, consumer)) {
+    const parameterValue = createNestedStackParameter(consumer, reference);
+    return resolveValue(consumer, parameterValue);
```

Comment:
Im missing something here, when will this call result in an actual recursion, the `parameterValue` is always defined on the on the consumer so this call will always result in:
```typescript
 if (producer === consumer) {
    return reference;
  }
```
https://github.com/aws/aws-cdk/pull/7187/files#diff-ce0a307ecd3c9f405ff4510a2516fd13R45

Reply:
If we have multiple levels of nesting (i.e. the producer is not a direct parent of the consumer), we will need to create a parameter in each parent until we eventually reach the producer.

Reply:
Since we don't change the consumer when we call the recursion we will not climb up the tree to get to the LSA.
I think that the reason this  pass tests is due to the reference being added on the nestedStack resource in the parent in `createNestedStackParameter`. Since the added assets triggers another round of `resolveReferences` while will resolve the added reference

Reply:
Yes I think you are correct that the recursive call will always stop immediately by returning the `parametetValue`. Let me clean this up.

Reply:
Fixed.

Reply:
Just to make sure I get it, we removed the recursion because we rely on `createNestedStackParameter ` to trigger another round of resolve?

Reply:
Here is what I find unintuitive in the current code. Assume we have the following stacks structure:
```
              A (Root)
              |
              B (Nested)
              |
              C (Nested)
```
Stack C references a resource in stack A, which requires wiring the parameter from A all the way down to C.  Since we don't recurse when creating the parameter on the consumer, the first round of `resolveReferences` will not create the required parameter on B (it will only create the parameter in C's template and the parameter on the respective  NestedStack resource in B). The required parameter on B will only be created on the subsequent round of `resolveReferences` which is only triggered due to `defineNestedStackAssets` returning true. This works, however, it's not intuitive. `defineNestedStackAssets` logic and name implies it will trigger only the resolution of asset references but its return result is relied upon to trigger the resolving of other references. In the above examples, B's parameters are created as a result of this side effect. For example one might see this code:
```typescript 
while (more) {
  resolveReferences(root);
  more = defineNestedStackAssets(nestedStacks);
}
```

And change it to:

```typescript
resolveReferences(root);
while (more) {
  more = defineNestedStackAssets(nestedStacks);
}
```

As the coupling between methods is not clear from their names or the `more` parameter name. 
Since this works due to the DFS order in which we process nestedStacks we can keep the code but maybe we should change the flag name to be more intuitive, e.g `moreResolveRequired`.

Reply:
It’s a good point. Let me see if I can improve.